java-util/src/main/java/io/druid/java/util/common/lifecycle/Lifecycle.java

```diff
@@ -241,59 +257,85 @@ public void addMaybeStartHandler(Handler handler) throws Exception
+    if (!startStopLock.tryLock()) {
+      // (*) This check is why the state should be changed before startStopLock.lock() in stop(). This check allows to
+      // spot wrong use of Lifecycle instead of entering deadlock, like https://github.com/druid-io/druid/issues/3579.
+      if (state.get().equals(State.STOP)) {
+        throw new ISE("Cannot add a handler in the process of Lifecycle stopping");
+      }
+      startStopLock.lock();
+    }
+    try {
+      if (state.get().equals(State.STOP)) {
+        throw new ISE("Cannot add a handler after the Lifecycle has stopped");
+      }
+      if (state.get().equals(State.RUNNING)) {
+        if (stage.compareTo(currStage) <= 0) {
           handler.start();
         }
       }
       handlers.get(stage).add(handler);
     }
+    finally {
+      startStopLock.unlock();
+    }
   }
 
   public void start() throws Exception
   {
-    synchronized (handlers) {
-      if (!started.compareAndSet(false, true)) {
+    startStopLock.lock();
+    try {
+      if (!state.get().equals(State.NOT_STARTED)) {
         throw new ISE("Already started");
       }
-      for (Stage stage : stagesOrdered()) {
-        currStage = stage;
-        for (Handler handler : handlers.get(stage)) {
+      if (!state.compareAndSet(State.NOT_STARTED, State.RUNNING)) {
+        throw new ISE("stop() is called concurrently with start()");
+      }
+      for (Map.Entry<Stage, ? extends List<Handler>> e : handlers.entrySet()) {
+        currStage = e.getKey();
+        for (Handler handler : e.getValue()) {
           handler.start();
         }
       }
     }
+    finally {
+      startStopLock.unlock();
+    }
   }
 
   public void stop()
   {
-    synchronized (handlers) {
-      if (!started.compareAndSet(true, false)) {
-        log.info("Already stopped and stop was called. Silently skipping");
-        return;
-      }
-      List<Exception> exceptions = Lists.newArrayList();
-
-      for (Stage stage : Lists.reverse(stagesOrdered())) {
-        final CopyOnWriteArrayList<Handler> stageHandlers = handlers.get(stage);
-        final ListIterator<Handler> iter = stageHandlers.listIterator(stageHandlers.size());
-        while (iter.hasPrevious()) {
-          final Handler handler = iter.previous();
+    // This CAS outside of a block guarded by startStopLock is the only reason why state is AtomicReference rather than
+    // a simple variable. State change before startStopLock.lock() is needed for the new state visibility during the
+    // check in addMaybeStartHandler() marked by (*).
+    if (!state.compareAndSet(State.RUNNING, State.STOP)) {
+      log.info("Already stopped and stop was called. Silently skipping");
+      return;
+    }
+    startStopLock.lock();
+    try {
+      RuntimeException thrown = null;
+      for (List<Handler> stageHandlers : handlers.descendingMap().values()) {
+        for (Handler handler : Lists.reverse(stageHandlers)) {
           try {
             handler.stop();
```

Comment:
currStage is checked only in some logic in `addMaybeStartHandler()` method, that is unreachable since `stop()` is called. So it's pointless. It wasn't done before in this method as well.